File: src/main/java/ru/mydesignstudio/database/metadata/extractor/extractors/udf/UdfExtractor.java
package ru.mydesignstudio.database.metadata.extractor.extractors.udf;

import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Component;
import ru.mydesignstudio.database.metadata.extractor.extractors.ExtractHelper;
import ru.mydesignstudio.database.metadata.extractor.extractors.TableMetadataAppender;
import ru.mydesignstudio.database.metadata.extractor.extractors.model.TableMetadata;
import ru.mydesignstudio.database.metadata.extractor.resource.StringResource;

@Component
public class UdfExtractor implements TableMetadataAppender {
  @StringResource("classpath:sql/extract_udfs.sql")
  private String extractQuery;

  @Autowired
  private ExtractHelper helper;

  @Override
  public void append(TableMetadata metadata, String schemaName, String tableName) {
    metadata.setUdfs(helper.extract(extractQuery, new Object[] { tableName, schemaName }, UdfModel.class));
  }
}

File: extractor-source/source-oracle/src/main/kotlin/ru/mydesignstudio/database/metadata/extractor/table/extractor/UdfExtractor.kt
package ru.mydesignstudio.database.metadata.extractor.table.extractor

import org.springframework.stereotype.Component
import ru.mydesignstudio.database.metadata.extractor.ExtractorBase
import ru.mydesignstudio.database.metadata.extractor.connection.DataSourceProvider
import ru.mydesignstudio.database.metadata.extractor.extract.helper.QueryProvider
import ru.mydesignstudio.database.metadata.extractor.extract.parameters.source.MetadataObject
import ru.mydesignstudio.database.metadata.extractor.extract.parameters.source.MetadataSchema
import ru.mydesignstudio.database.metadata.extractor.extract.parameters.source.MetadataSourceConnection
import ru.mydesignstudio.database.metadata.extractor.extract.result.UdfModel

@Component
class UdfExtractor (
        private val dataSourceProvider: DataSourceProvider,
        private val queryProvider: QueryProvider) : ExtractorBase() {
    fun extract(metadata: MetadataObject, schema: MetadataSchema, connection: MetadataSourceConnection): Collection<UdfModel> {
        val query = queryProvider.provide("classpath:sql/extract_udfs.sql")
        val dataSource = dataSourceProvider.provide(connection)
        return extract(dataSource, query, listOf(metadata.name, schema.name)) { rs ->
            UdfModel(
                    schemaName = getString(rs, "schema_name"),
                    functionName = getString(rs, "function_name"),
                    returnType = getString(rs, "return_type"),
                    arguments = getString(rs, "arguments")
            )
        }
    }
}